File: back/models/Achat.js
const mongoose = require('mongoose');


const AchatSchema = new mongoose.Schema({
  id_article: { type: mongoose.Schema.Types.ObjectId, ref: 'Article', required: false },
  id_fournisseur: { type: mongoose.Schema.Types.ObjectId, ref: 'fournisseur', required: false },
  id_utilisateur: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: false },
  prix_achatHT: { type: Number, required: false },
  date_achat: { type: Date, required: false },
  type_achat: { type: String, enum: ['Direct', 'Commandé'], required: false },
  validation_admin: { type: Boolean, required: false },
  TVA: { type: Number, required: false },
  prix_achatTTC: { type: Number, required: false },
  id_paiement: { type: mongoose.Schema.Types.ObjectId, ref: 'Paiement', required: false },
  quantité: { type: Number, required: false },
  numACHAT: { type: String, required: false },
  delai_livraison: { type: Number, required: false },
  id_document: { type: mongoose.Schema.Types.ObjectId, ref: 'Document', required: false },
}, { timestamps: true });

// Création du modèle
module.exports = mongoose.model('Achat', AchatSchema);

File: mobilef/FRENT/lib/models/achat.dart
class Purchase {
  final String? id;
  final String id_article;
  final String id_fournisseur;
  final double prix_achatHT;
  final double TVA;
  final int quantite;
  final double prix_achatTTC;
  final int? delai_livraison;
  final String type_achat;
  final DateTime date_achat;

  Purchase({
    this.id,
    required this.id_article,
    required this.id_fournisseur,
    required this.prix_achatHT,
    required this.TVA,
    required this.quantite,
    required this.prix_achatTTC,
    this.delai_livraison,
    required this.type_achat,
    required this.date_achat,
  });

  Map<String, dynamic> toJson() => {
        '_id': id,
        'id_article': id_article,
        'id_fournisseur': id_fournisseur,
        'prix_achatHT': prix_achatHT,
        'TVA': TVA,
        'quantite': quantite,
        'prix_achatTTC': prix_achatTTC,
        'delai_livraison': delai_livraison,
        'type_achat': type_achat,
        'date_achat': date_achat.toIso8601String(),
      };

  factory Purchase.fromJson(Map<String, dynamic> json) => Purchase(
        id: json['_id']?.toString(),
        id_article: json['id_article'].toString(),
        id_fournisseur: json['id_fournisseur'].toString(),
        prix_achatHT: double.parse(json['prix_achatHT'].toString()),
        TVA: double.parse(json['TVA'].toString()),
        quantite: int.parse(json['quantite'].toString()),
        prix_achatTTC: double.parse(json['prix_achatTTC'].toString()),
        delai_livraison: json['delai_livraison'] != null ? int.parse(json['delai_livraison'].toString()) : null,
        type_achat: json['type_achat'].toString(),
        date_achat: DateTime.parse(json['date_achat']),
      );

  static double calculateTTC(double prixHT, double tva, int quantite) {
    return prixHT * quantite * (1 + tva / 100);
  }
}
